lib/controllers/PermissionsController.ts

```diff
@@ -3,28 +3,108 @@ import BaseController from './BaseController';
+    const permission = PermissionsController.getPermissionGrant(request);
+
+    const result = await this.context.store.createDocument({
+      owner: request.aud,
+      schema: PERMISSION_GRANT_SCHEMA,
+      payload: permission,
+    });
+
+    return HubResponse.withObject(result);
   }
 
   async handleExecuteRequest(request: HubRequest): Promise<HubResponse> {
     throw new HubError(`${request.getAction()} handler not implemented.`, HttpStatus.NOT_IMPLEMENTED);
   }
 
   async handleReadRequest(request: HubRequest): Promise<HubResponse> {
-    throw new HubError(`${request.getAction()} handler not implemented.`, HttpStatus.NOT_IMPLEMENTED);
+    PermissionsController.validateSchema(request);
+
+    const results = await this.context.store.queryDocuments({
+      owner: request.aud,
+      schema: PERMISSION_GRANT_SCHEMA,
+    });
+
+    if (request.request && request.request.id) {
+      const id = request.request.id;
+      const match = results.filter(result => result.id === id);
+      if (match.length > 0) {
+        return HubResponse.withObject(match[0]);
+      }
+      return HubResponse.withError(new HubError('Permission not found', 404));
+    }
+    return HubResponse.withObjects(results);
   }
 
   async handleDeleteRequest(request: HubRequest): Promise<HubResponse> {
-    throw new HubError(`${request.getAction()} handler not implemented.`, HttpStatus.NOT_IMPLEMENTED);
+    PermissionsController.validateSchema(request);
+
+    if (!request.request || !request.request.id) {
+      throw new HubError('request.id is required', 400);
+    }
+
+    const id = request.request.id;
+
+    await this.context.store.deleteDocument({
+      id,
+      owner: request.aud,
+      schema: PERMISSION_GRANT_SCHEMA,
+    });
+
+    return HubResponse.withSuccess();
   }
 
   async handleUpdateRequest(request: HubRequest): Promise<HubResponse> {
-    throw new HubError(`${request.getAction()} handler not implemented.`, HttpStatus.NOT_IMPLEMENTED);
+    PermissionsController.validateSchema(request);
+    const permission = PermissionsController.getPermissionGrant(request);
+    if (!request.request || !request.request.id) {
+      throw new HubError('request.id is required', 400);
+    }
+
+    const id = request.request.id;
+
+    const result = await this.context.store.updateDocument({
+      id,
+      owner: request.aud,
+      schema: PERMISSION_GRANT_SCHEMA,
+      payload: permission,
+    });
+
+    return HubResponse.withObject(result);
+  }
+
+  // given a hub request, attempts to retrieve the PermissionGrant from it's payload
+  private static getPermissionGrant(request: HubRequest): PermissionGrant {
+    if (!request.payload) {
+      throw new HubError('request.payload required', 400);
+    }
+    const permission = request.payload.data as PermissionGrant;
+    if (!permission.owner ||
+        !permission.grantee ||
+        !permission.allow ||
+        !permission.object_type ||
+        !permission.created_by) {
```

Comment:
Is the `created_by` property of a grant mandatory? I suppose it's fine for now if it is. Later on we may want to allow this to be an optional property.

Reply:
This should be optional however is currently ignored.